### ISSUE TITLE
Reproduction of results.

### ISSUE THREAD

**Issue by nasosger** [ziqipang/RandAR#7]
Hello! 
Congrats for this nice work, and the CVPR acceptance, and also thank you for open-sourcing everything! 
I would like to ask some things regarding your implementation.

1. First of all, are the results from the paper reproducible with the given code? (e.g. with the 0.3B model). I see that you have a default different learning rate scheduler (you mention constant lr in the paper, with linear decay in the last epochs), so I guess if i change this i will have the exact same settings as in the paper. Did the cosine scheduler help at all?

2. What exactly is the raster-order counterpart which you compare with? You interleave the position tokens, but you do not shuffle the sequences, as far as i understand. So you get some extra computation from the position instruct tokens (reminds me of the "pause tokens" paper). Have you verified if this raster-order counterpart with position tokens performs any better than the vanilla auto-regressive model (e.g. LlamaGen baseline)? It would be interesting.

3. And finally, as far as i understood from the code, the position tokens (Q,K) are rotated twice by 2D-ROPE. You do rotate the learnable representations before interleaving the sequences, then  you rotate again the Q,K vector of these tokens with the exact same rotation (correct me if I am wrong :P).  I guess this works because the value vectors need to be rotated though.
Anyway, congrats again for this very nice work, and thank you in advance for any clarifications!

**Comment by ziqipang** [comment]
@nasosger Thanks for the nice words and interest!

1. I expected the cosine learning rate to help, but unfortunately it didn't (please check the table in the readme for detailed numbers). The 0.3B model performs slightly worse with cosine lr, and the 0.7B model performs similarly. 
2. Yes, we verified that such a counterpart performs similar with the original LLaMAGen. The number in the paper is better than LLaMAGen's original number because we applied linear CFG for both raster-order and our random-order models for a fair comparison.
3. This is a very good observation! (1) We rotate the values of the position instruction tokens, so that they can represent "positions" like (x, y). (2) the rotation inside the attention for QK is standard attention. 
 
I hope this clarifies and answers your concerns!

**Comment by a1600012888** [comment]
add more details:

regarding your second points:
> So you get some extra computation from the position instruct tokens (reminds me of the "pause tokens" paper).  
Yes, you are right. 
Also the fid improvement also comes from the learning rate decay in the last 50 epoch. (llamagen use constant lr throughout the training, for the numbers on the paper, we used constant for 250 epochs and linearly decay in the last 50 epoch).

> Have you verified if this raster-order counterpart with position tokens performs any better than the vanilla auto-regressive model (e.g. LlamaGen baseline)? It would be interesting.
I couldn't recall the details, according to my fuzzy we tried a comparison at an early age, using imagenet-128 res and the 343M sized model.  I tried to look it up in the wandb history and find these two experiments. It should be a llamagen and the raster-order counterpart (with interleaved position instruct token, and raster order). 
The image shows the interleaved format has better training loss. (its all around 250 epoch of training with the same learning rate). But I could recall the details of these two experiments, and it should be a apple to apple comparison (model size, hyperparam, data epoch, raster-order).

![Image](https://github.com/user-attachments/assets/7910792a-66b0-42f7-90de-664fcca57048)

**Comment by nasosger** [comment]
@ziqipang @a1600012888 
Thank you guys! I really find your project really interesting, I want to run some experiments to try some things out. That's why I am asking for details, since I don't have much computational resources, and I will be able to run a few experiments, with the ~110 M model ( unfortunately, doubling the context length hurts the GPU memory ). What about the augmentations, i see that you propose using the adm (2 crops), I am a bit afraid that this might lead to some overfitting though (since Imagenet is not a large dataset itself for this type of training).

**Comment by ziqipang** [comment]
@nasosger Thanks for the follow-up! From the perspective of generative models, the adm style of augmentation is widely used for ImageNet, so we think it is not a big problem. Since you are also training a 100M model, I think you don't need to worry about overfitting.

**Comment by nasosger** [comment]
@a1600012888 @ziqipang Thanks for the discussion guys! You were very helpful. I will let you know If I build anything interesting on top of your work.

**Comment by ziqipang** [comment]
Nice! Looking forward to that
